internal/ctrlflow/ctrlflow.go

```diff
@@ -0,0 +1,184 @@
+package ctrlflow
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"log"
+	mathrand "math/rand"
+	"strconv"
+	"strings"
+
+	"golang.org/x/tools/go/ast/astutil"
+	"golang.org/x/tools/go/ssa"
+	ah "mvdan.cc/garble/internal/asthelper"
+	"mvdan.cc/garble/internal/ssa2ast"
+)
+
+const (
+	mergedFileName = "GARBLE_controlflow.go"
+	directiveName  = "//garble:controlflow"
+	importPrefix   = "___garble_import"
+
+	defaultSplit  = 0
+	defaultJunk   = 0
+	defaultPasses = 1
+)
+
+type directiveParamMap map[string]string
+
+func (m directiveParamMap) GetInt(name string, def int) int {
+	rawVal, ok := m[name]
+	if !ok {
+		return def
+	}
+
+	val, err := strconv.Atoi(rawVal)
+	if err != nil {
+		panic(fmt.Errorf("invalid flag %s format: %v", name, err))
+	}
+	return val
+}
+
+// parseDirective parses a directive string and returns a map of directive parameters.
+// Each parameter should be in the form "key=value" or "key"
+func parseDirective(directive string) (directiveParamMap, bool) {
+	if !strings.HasPrefix(directive, directiveName) {
+		return nil, false
+	}
+
+	fields := strings.Fields(directive)
+	if len(fields) <= 1 {
+		return nil, true
+	}
+	m := make(map[string]string)
+	for _, v := range fields[1:] {
+		params := strings.SplitN(v, "=", 2)
+		if len(params) == 2 {
+			m[params[0]] = params[1]
+		} else {
+			m[params[0]] = ""
+		}
+	}
+	return m, true
+}
+
+// Obfuscate obfuscates control flow of all functions with directive using control flattening.
+// All obfuscated functions are removed from the original file and moved to the new one.
+// Obfuscation can be customized by passing parameters from the directive, example:
+//
+// //garble:controlflow passes=1 junk=0 split=0
+// func someMethod() {}
+//
+// passes - controls number of passes of control flow flattening. Have exponential complexity and more than 3 passes are not recommended in most cases.
+// junk - controls how many junk jumps are added. It does not affect final binary by itself, but together with flattening linearly increases complexity.
+// split - controls number of times largest block must be splitted. Together with flattening improves obfuscation of long blocks without branches.
```

Comment:
Does this mean that control flow obfuscation is opt-in on a per-function basis? I think long-term that's not the best strategy, because if it works well and the performance impact isn't that bad with some default parameters, I imagine many users would want it to be turned on for all functions.

Perhaps we should turn it on by default once it's been tested and benchmarked well enough, and still allow users to add the directives to tweak the parameters - or turn it off entirely for a particular function.

Reply:
I would also suggest using longer names which are more descriptive. For example: `flatten_passes`,  `junk_jumps`, and `block_splits`.

Reply:
it's hard for me to imagine that feature being enabled globally, for example converting an iterator over a map can uncontrollably kill the entire performance.

i think it is more feasible to allow users to specify default values per-package (now these are defaultBlockSplits, defaultJunkJumps, defaultFlattenPasses constants)

Reply:
Gotcha. You could always allow the same directives to apply to an entire file by being placed before the `package` line, similar to `//go:build`. I think that's probably more feasible than per-package, because per-package we don't really have a good place for the directive or any config file to go.

Reply:
I think that's out of this pullquest. Other nits fixed - merge?

Reply:
If you mean the per-file setting, of course - I was just following with the thinking outloud :)

Reply:
i didn't write in a comment: I agree with per-file idea :)